src/server/saved_objects/service/lib/priority_collection.ts

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+interface PriorityCollectionEntry<T> {
+  priority: number;
+  value: T;
+}
+
+export class PriorityCollection<T> {
+  private readonly array: Array<PriorityCollectionEntry<T>> = [];
+
+  public add(priority: number, value: T) {
+    let i = 0;
+    while (i < this.array.length) {
```

Comment:
Let's not use `while` loops for iterating arrays. They provide no protections against infinite loops and are often impacted by off-by-one errors, particularly when developers unfamiliar with the code start making changes in the future.

If you still want procedural, perhaps something like:

```js
const foundIndex = this.array.findIndex(current => {
  // throw if priority === current.priority
  return priority < current.priority;
});

const spliceFrom = foundIndex > -1 ? foundIndex : this.array.length - 1;

// this.array.splice(spliceFrom, ...)
```

or if you want something a bit more functional:

```js
this.array = this.array.reduce((arr, current, index) => {
  // throw if priority === current.priority
  if (priority < current.priority) {
    // arr.splice(index, ...)
  }
  return arr;
}, []);
```

or my favorite, get properly ecmascripted up in here:

```js
for (const [index, value] of this.array.entries()) {
  // throw if priority === current.priority
  if (priority < current.priority) {
    // this.array.splice(index, ...)
    break;
  }
}

```